lib/sql.js

```diff
@@ -1092,6 +1096,15 @@ SQLConnector.prototype._buildWhere = function(model, where) {
       }
       // The value is not an array, fall back to regular fields
     }
+
+    if (key === 'where') {
+      // business as usual if the model has a property named 'where'
+      if (props[key]) {
+        continue;
+      } else {
+        throw new Error('Filter object detected. Please use a where object instead.');
+      }
+    }
     var p = props[key];
     if (p == null) {
       // Unknown property, ignore it
```

Comment:
IMO, the change proposed above is not really fixing the problem. 

For example, consider a `Product` model with a `name` property, the developer making a typo in their query and calling `Product.destroyAll({names: 'Pen'})`. Right now, the entire SQL table is deleted. IMO, the correct behavior is that no records should have been deleted.

I think the root cause of this whole issue lies here.

```js
    var p = props[key];
     if (p == null) {
       // Unknown property, ignore it
       debug('Unknown property %s is skipped for model %s', key, model);
       continue;
     }
```

Instead of ignoring unknown properties, we need to construct a WHERE statement that is not going to match any record, or perhaps throw an error as you proposed above (but please use a different error message to better describe the new intent).

Reply:
Good catch! Thank you for illustrating what is IMO another big issue from the original ticket (among others). While I'm working on that, I want to keep what I have because it ensures rejection of `filter` objects. Do you have suggestions on how I can improve that (test case is [here](https://github.com/strongloop/loopback-datasource-juggler/pull/1597/files#diff-404450f6fd92e7f3f2031232213c07c5R1705)?

Reply:
How about the following?

```diff
 var p = props[key];
 if (p == null) {
-  // Unknown property, ignore it
-  debug('Unknown property %s is skipped for model %s', key, model);
-   continue;
+  throw new Error(
+    `Unknown property ${key} used in a "where" condition for model ${model}`
+  );
}
```